### ISSUE TITLE
Feature request: Support RAID 

### ISSUE THREAD

**Issue by ms178** [calamares/calamares#987]
#### Submission type

  - [ ] Bug report
  - [X] Feature Request

#### Info regarding which version of Calamares is used, which Distribution

> Manjaro 18.0 Beta 2

#### Provide information on how the disks are set up, in detail, with full logs of commands issued 

> RAID-0 as configured in the Intel Option ROM of my BIOS. It is on a X58 mainboard with ICH10R.

#### What do you expect to have happen when Calamares installs?

> Seeing my Intel ICH10R ( = fake RAID / software RAID solution which is provided on Intel mainboard chipsets) RAID-0 partitions and having the possibility to install the OS on one of them.

#### Describe the issue you encountered

> Currently, the installer only gets to see both HDDs which are physically installed on the system as SATA devices but not the already RAIDed parititions. Hence using these is not possible currently.

#### Steps to reproduce the problem

> Using the mentioned distro ISO and try to install it on a RAID-0 partition on a Intel ICH10R (or possibly any other Intel RAID) system. 

#### Include the installation.log (usually ~/Calamares/Calamares/Calamares.log, of the user Calamares runs as):

> n/a since installation is not possible due to not supporting RAID-0

**Comment by ms178** [comment]
By the way, the same issue was already mentioned in #794 and #806 - but as it still persits, this is requested again specifically as a feature request and not a bug this time.

**Comment by adriaandegroot** [comment]
Good of you to refer to previous issues that were filed. I think part of the problem is illustrated in #794: I don't have that kind of hardware, and have no idea what it's supposed to look like when it is working. So it's a good feature request, but it's going to need someone to specifically work on it (and submit a PR which looks at least plausible, as opposed to just hacking in some extra calls to mdadm or whatever).

It's possible -- likely even -- that there is also an issue in KPMCore: does KDE partition manager see this kind of RAID'ed disks?

**Comment by kkofler** [comment]
Uh, actually, the Linux kernel knows 2 kinds of RAID setups:
* `mdraid`: pure software RAID set up with `mdadm`,
* `dmraid`: software/fake RAID set up by a "hardware RAID" controller, detected by `device-mapper`.

(True hardware RAID is entirely transparent to the kernel to begin with, so should need no special setup.)

As I understand it, what is requested here is support for `dmraid`, not `mdraid`. But `mdraid` is not yet supported either.

**Comment by ms178** [comment]
I might add that the Ubuntu installer gets this right and recognizes the Intel Fake-Raid partitions correctly. So Ubuntu is one of the few distros which I could use at the moment on my Workstation. Maybe a look into their sources could be a source for inspiration? By the way, mdraid is requested by me here as it is the suggested way even from Intel, see this White Paper from 2011: https://www.intel.com/content/dam/www/public/us/en/documents/white-papers/rst-linux-paper.pdf

**Comment by adriaandegroot** [comment]
@Vfreeze31 as a workaround, assuming your interest is mostly in RAID-0 for the gaming bits, is to install Manjaro first to a custom partitioning scheme with just space for / (and possibly swap), then once the system is installed, set up the RAID bits by hand.

That said, Calamares master (will be 3.2.3) has some RAID features from Caio merged in. That may help. For 3.2.3 there is nothing particular planned for RAID support -- I consider it pretty low-priority compared to other fixes and improvements in flexibility of Calamares.

**Comment by cjlcarvalho** [comment]
All the features for full support for software RAID (i.e. mdraid) are **almost** completed in kpmcore raid-support branch, but I have to finish implementing some particular features and fixing some processes like resizing and deleting. I hope to get a time during the next months to code it.

**Comment by gwolf2u** [comment]
Is this still under consideration?
I know there are workarounds, yet for average Joe all are rather hard to understand and needs some basic knowledge of partitioning through command line and so

**Comment by adriaandegroot** [comment]
@gwolf2u As long as the issue is open, it's still under consideration. I think @cjlcarvalho got all the pieces lined up. What it needs is for someone who cares about Linux RAID and has the means to test if it works, to do a PR to finish it off. Or, heck, documenting the workarounds would probably help as well. For me, it's just lower-priority-than-the-things-that-need-doing-now.

**Comment by brettroyerr** [comment]
I'm building a new PC with a similar situation to @Vfreeze31. I am excitedly awaiting this feature in the next Manjaro installer. Any documentation to the possible workarounds would be greatly appreciated as that should hold me off in the mean time.

**Comment by gwolf2u** [comment]
What I did to install previously:
1. Create the RAID in BIOS (FakeRAID in my case)
2. Boot live disk and open terminal then run setup command (basically installing via command line)
3. Created partitions using gparted (this needs to be installed) in order to make it list the partitions (made 1GB efi, 50GB for boot and rest for home)
4. Mounted all partitions from terminal setup 
5. Chosen on terminal setup what type of install I want and left it install
6. Installed grub bootloader (when asked - resulted in and error yet package got installed so will use it later) and chrooted to installation 
7. Here I do some config
- set mdadm.conf to list RAID
`mdadm --examine --scan > /etc/mdadm.conf`
- set mkinitcpio
`sudo nano /etc/mkinitcpio.conf`
- added /sbin/mdmon to the BINARIES section (mdmon monitors external metadata and will be automatically executed by udev rules)
`BINARIES="/sbin/mdmon"`
- added mdadm_udev to the HOOKS section between block and filesystems (mdadm_udev is an installation hook that includes mdadm and the udev rules responsible for RAID assembly in the initramfs)
`HOOKS="base udev autodetect modconf block mdadm_udev filesystems keyboard fsck"`
- saved file
- generated initramfs
`mkinitcpio -P`
- set grub (in my case since I use EFI I ran below commands - seek Google for BIOS ones)
`grub-install --target=x86_64-efi --efi-directory=/boot/efi --bootloader-id=GRUB`
`grub-mkconfig -o /boot/grub/grub.cfg`

8. Reboot and enjoy :smile: 

There might be another way around, yet this is what I used
So as you can see this is rather tricky to install 
A GUI installation is much more appropriate if possible 
Linux and RAID duh :grin:

**Comment by andrew-klassen** [comment]
The Architect installer recently got RAID support, I wrote it. I'm not sure RAID would be appropriate for the Calamares installer though, because its a more advanced feature.

**Comment by hhannine** [comment]
> 
> 
> The Architect installer recently got RAID support, I wrote it. I'm not sure RAID would be appropriate for the Calamares installer though, because its a more advanced feature.

My laptop came with a BIOS raid0 from the factory and I'm not advanced enough user to handle the installation manually so I would like to make the argument that there are devices and users that need (guided) support for RAID in the graphical installer.

**Comment by abucodonosor** [comment]
First of all the FaikRaid from BIOS, eg: AMD ( which isn't even supported in Linux )
and Intel is pure and utter s*it. Simply do not use that in Linux or in general.

That is not the Windows installer is how RAID works.
You'll see the same on HW RAID or Linux/Windows SW RAID.

FakeRAID need vendors driver support which is, believe it or not, closed source,
or partially available source, not support in any way in Linux upstream and with restrictive download & redistribution licenses.

While Intel or dmraid somewhat works the AMD FakeRaid is not supported at all.

Also please note using FakeRaid is slower anyway but the point of not using it
is bc is depends on your HW. Container information is stored in the hardware/firmware
meaning, the worse things can happen if your HW dies are you to lose all your DATA.

And not only when your HW dies but also on any disc issue. You have to pray the BIOS
vendors did the right thing and you can rebuild the array.

In other words even in Windows use the SW RAID they offer.

**Comment by ms178** [comment]
The argument is not about if that is a good solution or if the user should consider other options, it is about if Calamares could technically support it and could provide it as an option to its users if they want to use it. Today they need to go to another Linux distribution which provide this option (e.g. Ubuntu/OpenSuse).

**Comment by abucodonosor** [comment]
Sure supporting Linux RAID.

Again FakeRAID may not be possible at all in Linux for some HW & Vendors.

**Comment by KIC** [comment]
does architect meanwhile support software raid? mdadm is actually not depend one any bios fake raid. just provide 2 similar disks

**Comment by andrew-klassen** [comment]
> does architect meanwhile support software raid? mdadm is actually not depend one any bios fake raid. just provide 2 similar disks

Yes, the architect installer can provide you with RAID 0,1,5,6, or 10 through mdadm as an extra volume or bootable. The architect installer was also designed with nested configurations in mind. For an example, deploying LVM or a LUKS volume on top of an mdadm array does not break the installer. This was one aspect I did not like about the server installers from Ubuntu and Debian.

**Comment by andrew-klassen** [comment]
I think you are right about the calamares installer not fully supporting RAID, but I verified an hour ago that you can get the calamares installer to work by doing parts of the installation manually from the command line. 

2 years ago, I would have recommended that you ditch calamares all together and use the architect installer because it is much better suited for this type of install, but now the project is no longer maintained. However, before I added RAID support to Manjaro architect, I made a video of how to do the process manually. The content is still relevant.

[https://www.youtube.com/watch?v=eRxUZxUnJq0](https://www.youtube.com/watch?v=eRxUZxUnJq0)

Below is the video notes.

[https://github.com/andrew-klassen/manjaro_architect_uefi_raid](https://github.com/andrew-klassen/manjaro_architect_uefi_raid)

When I tested an hour ago, here is what I did at a high level. I tested by creating a bootable RAID 0 array across 2 virtual disks.

1. Shred/zero fill all disks that you plan to use as RAID members and then reboot into the manjaro installation disk.
2. Drop to the terminal in the installer. Install the mdadm package and then build your partitioning scheme+RAID array in the terminal. When I did it, I noticed that manajaro build my raid array using the incorrect device name (/dev/md127). I manually disassembled and re-assembled the array and it fixed the issue.

`sudo mdadm --stop /dev/md127`
`sudo mdadm --assemble --verbose /dev/md0 /dev/sda1 /dev/sdb1`

3. Install your filesystem on the RAID array, but do not mount it.
4. Start the install and go through it normally until you get to the disk partitioning part.
5. Select the manual method and you should see your RAID array. Edit it and adjust the mount point.
6. Finish the rest of the Calamares installation but do not reboot the system afterwards. You will have to do some additional work to make the array bootable.
7. Add the hook and binary to the mkinicpio.conf
8. Re-build the inital ramdisk in a chroot root jail (refer to my video notes on how to do this). You will want to use the manjaro-chroot command line utility for this, its available in the installer.
9. Add the RAID scan details in the mdadm.conf.
10. Then install and update grub.

You might want to watch the entire video before attempting this because the process is almost as involved as doing an Arch installation. But by doing this, you get the standard Manjaro software.

**Comment by ms178** [comment]
Thanks a lot for the detailed writeup, Andrew. The last four points seem to be very involved, while I am not a Linux novice anymore I'd very much prefer if  it would be just as easy as using the Ubuntu installer. :) But thanks nevertheless, I might be willing to motivate me somewhen in the future to learn how to execute the last steps as there are some things I can only use with Calamares-based distros right now and while I do have quite a few different storage devices installed in my system that provide an easy workaround for this issue, it is the RAID-0 array that is dedicated for my Linux tinkering.

**Comment by taojoe** [comment]
3.2.42 has already released, so is this feature supported?

**Comment by adriaandegroot** [comment]
@taojoe No; the issue is still open. The milestone names are suggestive, but do not correspond one-to-one with actual releases (I should stop putting "I intend to work on this" on milestones and instead just put issues on the milestone when closed). A PR from someone who understands RAID setup would go a long long way.

**Comment by taojoe** [comment]
@adriaandegroot  is there a plan for this feature?

**Comment by adriaandegroot** [comment]
Removed from milestone to drop the suggestion that it is under consideration. It's not high on my priority list. PRs can change that.

**Comment by rayzorben** [comment]
Would still like to see this, another new laptop user here that came with RAID enabled (Intel VROC). I can get it working in Arch, it works in OpenSUSE, but any distro using Calamares including my favorite (CachyOS) results in an error even though I can select it as an install target.

**Comment by rayzorben** [comment]
Here is the error I am getting. It simply fails to install https://termbin.com/emmu

```
2024-08-15 - 13:18:58 [1]: void Calamares::ViewManager::onInstallationFailed(const QString&, const QString&)
2024-08-15 - 13:18:58 [1]:     ERROR: Installation failed: "The installer failed to create partition on disk 'md126'." 
2024-08-15 - 13:18:58 [6]:     .. - message: "The installer failed to create partition on disk 'md126'." 
2024-08-15 - 13:18:58 [6]:     .. - details: 
Create a new partition (2.88 TiB, luks) on ‘/dev/md126’


Job: Create new partition on device ‘/dev/md126’


Command: sfdisk --force --append /dev/md126

Failed to add partition ‘New Partition’ to device ‘/dev/md126’.

Failed to add partition ‘New Partition’ to device ‘/dev/md126’.


2024-08-15 - 13:18:58 [6]: void Calamares::ViewManager::onInstallationFailed(const QString&, const QString&)
2024-08-15 - 13:18:58 [6]:     Calamares will quit when the dialog closes. 
2024-08-15 - 13:18:58 [6]: DEBUG (Qt): QML Component (default slideshow) deactivated
2024-08-15 - 13:18:58 [6]: void Config::doNotify(bool, bool)
2024-08-15 - 13:18:58 [6]:     Notification not sent; completion: failed 
```

**Comment by Franz-Heinz** [comment]
I guess this still isn't a development priority?

I'm trying to install TUXEDO OS will full system encryption on a software RAID 1 but can't figure out, how to make it work. The installer just shows me the hardware devices (2 SSDs) but not the raid devices. So I'm not getting even near to install the OS.

Like suggested in other comments, the Ubuntu installer at least seems to be able to use faik-RAID with the build in controller after initialisation in BIOS/EFI. But like already explained, faik-RAID is nothing one would possibly want to use for important data.

**Comment by rayzorben** [comment]
> I guess this still isn't a development priority?
> 
> I'm trying to install TUXEDO OS will full system encryption on a software RAID 1 but can't figure out, how to make it work. The installer just shows me the hardware devices (2 SSDs) but not the raid devices. So I'm not getting even near to install the OS.
> 
> Like suggested in other comments, the Ubuntu installer at least seems to be able to use faik-RAID with the build in controller after initialisation in BIOS/EFI. But like already explained, faik-RAID is nothing one would possibly want to use for important data.

Can you open up a terminal in the installer as root and run `mdadm --assemble --scan`

If that works that it now shows your raid drives, prior to rebooting you will need to run the following, ensuring that `/mnt/etc` is where your new OS drive is, not the installer. `mdadm --detail --scan >> /mnt/etc/mdadm.conf`
